src/core/data/ions/particle_initializers/maxwellian_particle_initializer.h

```diff
@@ -95,317 +83,252 @@ namespace core
 
-            for (std::uint32_t ix = ix0; ix < ix1; ++ix)
-            {
-                for (std::uint32_t iy = iy0; iy < iy1; ++iy)
-                {
-                    double n; // cell centered density
-                    std::array<double, 3> particleVelocity;
-                    std::array<std::array<double, 3>, 3> basis;
-
-                    // get the coordinate of the current cell
-                    auto coord = layout.cellCenteredCoordinates(ix, iy);
-                    auto x     = coord[0];
-                    auto y     = coord[1];
-
-                    // now get density, velocity and thermal speed values
-                    n         = density_(x, y);
-                    auto Vx   = bulkVelocity_[0](x, y);
-                    auto Vy   = bulkVelocity_[1](x, y);
-                    auto Vz   = bulkVelocity_[2](x, y);
-                    auto Vthx = thermalVelocity_[0](x, y);
-                    auto Vthy = thermalVelocity_[1](x, y);
-                    auto Vthz = thermalVelocity_[2](x, y);
-
-                    // weight for all particles in this cell
-                    auto cellWeight = n * cellVolume / nbrParticlePerCell_;
-
-                    ParticleDeltaDistribution randPosX;
-                    ParticleDeltaDistribution randPosY;
+    MaxwellianInitFunctions const fns{density_,       bulkVelocity_, thermalVelocity_,
+                                      magneticField_, basis_,        x};
 
-                    if (basis_ == Basis::Magnetic)
-                    {
-                        auto Bx = magneticField_[0](x, y);
-                        auto By = magneticField_[1](x, y);
-                        auto Bz = magneticField_[2](x, y);
+    auto const [n, V, Vth] = fns();
+    auto const cellVolume  = layout.cellVolume();
+    auto generator         = getRNG(rngSeed_);
+    std::size_t cell_idx   = 0;
 
-                        localMagneticBasis({Bx, By, Bz}, basis);
-                    }
+    for (std::uint32_t ix = ix0; ix < ix1; ++ix)
+    {
+        auto const cellWeight   = n[cell_idx] * cellVolume / nbrParticlePerCell_;
+        auto const AMRCellIndex = layout.localToAMR(Point{ix});
+        ParticleDeltaDistribution randPosX;
+        std::array<double, 3> particleVelocity;
+        std::array<std::array<double, 3>, 3> basis;
 
+        if (basis_ == Basis::Magnetic)
+        {
+            auto const B = fns.B();
+            localMagneticBasis({B[0][cell_idx], B[1][cell_idx], B[2][cell_idx]}, basis);
+        }
 
-                    for (std::uint32_t ipart = 0; ipart < nbrParticlePerCell_; ++ipart)
-                    {
-                        maxwellianVelocity({Vx, Vy, Vz}, {Vthx, Vthy, Vthz}, generator,
-                                           particleVelocity);
+        for (std::uint32_t ipart = 0; ipart < nbrParticlePerCell_; ++ipart)
+        {
+            maxwellianVelocity({V[0][cell_idx], V[1][cell_idx], V[2][cell_idx]},
```

Comment:
what is "not cool" with this new form of the code, is that this functions is bound to index cells *the same way* input function values where done... but nothing guarantees it will be the case... if this function starts looping over cells in some other way it will stop working. How could we make sure the ordering stays the same?
Maybe by having a third party object/function that provide the 1D flat cell index from multidimensional index and that object/function would be used by both these methods and the init function initialization code?

Reply:
I believe this has been resolved, the comment is not appearing as "outdated" tho as the surrounding code was changed and not the line which the comment is attached to.